### PR TITLE
Upgrade and configure support folders

### LINKED ISSUE #7497: vic-machine upgrade supports inventory folders
vic-machine has been updated to support folders for create and delete.  The support of folders needs to be extended to upgrade.

This is primarily needed for VCHs deployed prior to 1.4 and the folder support.  The following approach can be utilized:
- After successful upgrade of the appliance VM
- Determine VCH parent folder
- if not the vm folder, then no additional processing - upgrade complete
- create a uniquely named folder (not the same name as appliance: i.e. `/vm/temp-vch-001`)
- move the VCH appliance and all VCH containerVMs into newly created folder
- rename folder to match the VCH appliance name
- upgrade complete

- [ ] update upgrade logic
- [ ] ensure backward compatability
- [ ] update / add to ci upgrade tests

### PR DESCRIPTION
**Pull request by cgtexmex**
Upgrade will now update the inventory for the VCH by
moving the VCH and containerVMs into a folder with the
same name as the VCH.  This will match the functionality
provided in create.

Upgrade also has a debug flag allowing for more verbose
output from the cli.

Configure will not make any inventory changes.

Fixes #7497

### CODE REVIEW

File: lib/install/validate/config_to_data.go [hunk 1]
```diff
@@ -107,7 +107,7 @@ func setApplianceResources(op trace.Operation, vm *vm.VirtualMachine, d *data.Da
 // setVCHResources will populate the configuration data based on the deployed VCH config
 func setVCHResources(op trace.Operation, vch *object.ResourcePool, d *data.Data) error {
 	var p mo.ResourcePool
-	ps := []string{"config.cpuAllocation", "config.memoryAllocation"}
+	ps := []string{"config"}
```
Comment (zjs) [inline lib/install/validate/config_to_data.go]: What's the purpose of this change?

File: lib/install/management/configure.go [hunk 2]
```diff
@@ -98,49 +102,146 @@ func (d *Dispatcher) Configure(vch *vm.VirtualMachine, conf *config.VirtualConta
...
+		}
+	}
 
 	// If successful try to grant permissions to the ops-user
-	if err == nil && conf.ShouldGrantPerms() {
-		err = opsuser.GrantOpsUserPerms(d.op, d.session, conf)
+	if conf.ShouldGrantPerms() {
+		// TODO: In a DRS environment that's not been upgraded to 1.4
+		// this will attempt to apply permissions to the VmFolder. Does
+		// that work & do we really want that?
```
Comment (zjs) [inline lib/install/management/configure.go]: That _seems_ desirable, but we should probably have a robot test to make sure it works. (I'm not sure we have CI tests for the permissions functionality, but now that CI is on VC, it should be possible to add one.)

cc @lcastellano

File: lib/install/validate/config_to_data.go (hunk 1 above)
Comment (cgtexmex) [reply lib/install/validate/config_to_data.go]: Achieves the same result w/less characters -- which isn't a great reason...happy to pull it out..

FYI: plan on opening a pr against `govmomi` implementing this functionality via `object.ResourcePool`...(and a few others) then a pr against vic-machine to utilize....

File: lib/install/management/configure.go (hunk 2 above)
Comment (cgtexmex) [reply lib/install/management/configure.go]: I _assumed_ the `VmFolder` -- which is datacenter wide -- was to broad and we'd not want to do it there.  But I'm not fluent on the permission code either, so I might be all wrong...

File: lib/install/validate/config_to_data.go (hunk 1 above)
Comment (zjs) [reply lib/install/validate/config_to_data.go]: I think there are a lot of `config.*` properties other than `config.cpuAllocation` and `config.memoryAllocation`, aren't there?

Comment (cgtexmex) [reply lib/install/validate/config_to_data.go]: Not really - in addition to those two there are:
- changeVersion: string 
- entity: moref
- lastModified: dateTime

File: lib/install/management/configure.go [hunk 3]
```diff
@@ -53,43 +55,45 @@ var (
...
-		err = errors.Errorf("Failed to reconfigure resources: %s", err)
-		return err
-	}
-
-	defer func() {
-		if err != nil {
-			d.rollbackResourceSettings(conf.Name, settings)
+	// Resource Pools not available in a DRS Disabled environment, so only attempt an update
+	// if DRS is Enabled and this is a configure action.
+	if d.session.DRSEnabled != nil && *d.session.DRSEnabled && d.Action == ConfigureAction {
```
Comment (hickeng) [inline lib/install/management/configure.go]: Should we be rolling back resource changes for a rollback action as well?

File: lib/install/management/configure.go [hunk 4]
```diff
@@ -98,49 +102,150 @@ func (d *Dispatcher) Configure(vch *vm.VirtualMachine, conf *config.VirtualConta
...
+// If the VCH resides in the VMFolder then a new folder will
+// be created and the VCH and all Containers will be moved into the
+// new folder.  If the VCH already resides in a folder other than
+// the VMFolder, then no action will be taken.
+//
+// vApp support planned: https://github.com/vmware/vic/issues/7670
+func (d *Dispatcher) updateInventory(name string) error {
+	defer trace.End(trace.Begin("", d.op))
+
+	appName := d.appliance.Name()
```
Comment (hickeng) [inline lib/install/management/configure.go]: Should this be `ObjectName()`? I notice feature/robo has changed `vm.Name` to `vm.ObjectName` in most locations.

File: lib/install/management/configure.go [hunk 5]
```diff
@@ -98,49 +102,150 @@ func (d *Dispatcher) Configure(vch *vm.VirtualMachine, conf *config.VirtualConta
...
+	// vApp support planned: https://github.com/vmware/vic/issues/7670
+	if currentFolder.Reference() != d.session.VMFolder.Reference() || d.vchPool.Reference().Type == "VirtualApp" {
+		d.op.Debugf("No inventory update completed for %s", name)
+		return nil
+	}
 
-		d.deleteUpgradeImages(ds, settings)
-		d.retryDeleteSnapshotByRef(newSnapshotRef, snapshotName, conf.Name)
+	// concatenate w/vch name so we can identify if failure
+	tempName := fmt.Sprintf("%s-%s", name, uid.New().Truncate().String())
```
Comment (hickeng) [inline lib/install/management/configure.go]: We should ensure that this doesn't violate the name length limits. It may be possible to use the name convention logic to do so (`lib/portlayer/util/util.go`).

File: lib/install/management/configure.go [hunk 6]
```diff
@@ -98,49 +102,150 @@ func (d *Dispatcher) Configure(vch *vm.VirtualMachine, conf *config.VirtualConta
...
+		d.op.Errorf("Unable to locate children during inventory update: %s", err)
 		return err
 	}
 
-	// compatible with old version's upgrade snapshot name
-	if oldSnapshot != nil && (vm.IsConfigureSnapshot(oldSnapshot, ConfigurePrefix) || vm.IsConfigureSnapshot(oldSnapshot, UpgradePrefix)) {
-		d.retryDeleteSnapshotByRef(&oldSnapshot.Snapshot, oldSnapshot.Name, conf.Name)
+	var vicVMs []types.ManagedObjectReference
+	for _, c := range children {
+		// This will make a vSphere call for each VM in the RP - this might
```
Comment (hickeng) [inline lib/install/management/configure.go]: I'm not a fan of `O(n)` here but it's not likely to be an issue. If reworking this in the future then moving to a call to get the combined results makes sense, i.e. not using the property collector `RetrieveOne` method.

File: lib/install/management/configure.go [hunk 7]
```diff
@@ -98,49 +102,150 @@ func (d *Dispatcher) Configure(vch *vm.VirtualMachine, conf *config.VirtualConta
...
+		d.op.Debugf("Children(%d) moved into folder", len(vicVMs))
+	}
+	_, err = tasks.WaitForResult(d.op, func(ctx context.Context) (tasks.Task, error) {
+		return target.Rename(d.op, d.vmPathName)
+	})
+	if err != nil {
+		// Much like folder.MoveInto this will not impact the ability for vic to operate, so
+		// only a warning is issued.
+		d.op.Warnf("Unable to rename folder(%s) to %s during inventory update: %s", tempName, name, err)
+		d.op.Warnf("Folder naming will not interfere with VCH operation, rename via VC if desired")
```
Comment (hickeng) [inline lib/install/management/configure.go]: We don't make any assumptions about folder name?

File: lib/install/management/configure.go [hunk 8]
```diff
@@ -205,54 +311,50 @@ func (d *Dispatcher) Rollback(vch *vm.VirtualMachine, conf *config.VirtualContai
...
+			return err
 		}
+		return nil
 	}
 
-	if soap.IsVimFault(err) {
-		if _, ok := soap.ToVimFault(err).(*types.SystemError); ok {
-			return true
+	// retry on any general SystemError
+	isSystemError := func(err error) bool {
```
Comment (hickeng) [inline lib/install/management/configure.go]: should we retry on the transient "TaskInProgress" or similar errors as well?

File: lib/install/management/create.go [hunk 9]
```diff
@@ -116,11 +116,11 @@ func (d *Dispatcher) createPool(conf *config.VirtualContainerHostConfigSpec, set
 	return nil
 }
 
-func (d *Dispatcher) uploadImages(files map[string]string) error {
+func (d *Dispatcher) uploadISOs(files map[string]string) error {
 	defer trace.End(trace.Begin("", d.op))
 
 	// upload the images
-	d.op.Info("Uploading images for container")
+	d.op.Info("Uploading vic iso images")
```
Comment (hickeng) [inline lib/install/management/create.go]: lower case `vic` - probably some sort of UX violation. Don't know that we need it there anyway. `Uploading ISO images` seems sufficient given this is vic-machine.

File: lib/install/management/dispatcher.go [hunk 10]
```diff
@@ -75,18 +120,16 @@ var diagnosticLogs = make(map[string]*diagnosticLog)
...
 	isVC := s.IsVC()
 	e := &Dispatcher{
+		Action:  action,
 		session: s,
 		op:      trace.FromContext(ctx, "Dispatcher"),
 		isVC:    isVC,
 		force:   force,
 	}
-	if conf != nil {
-		e.InitDiagnosticLogsFromConf(conf)
```
Comment (hickeng) [inline lib/install/management/dispatcher.go]: Having moved this to create specifically have we lost any functionality from the other actions?

File: lib/install/management/finder.go [hunk 11]
```diff
@@ -60,10 +59,10 @@ func (d *Dispatcher) NewVCHFromID(id string) (*vm.VirtualMachine, error) {
 		d.op.Errorf("Failed to find VM %q: %s", moref, err)
 		return nil, err
 	}
-	vmm = vm.NewVirtualMachine(d.op, d.session, ovm.Reference())
+	d.appliance = vm.NewVirtualMachine(d.op, d.session, ovm.Reference())
 
 	// check if it's VCH
-	if ok, err = d.isVCH(vmm); err != nil {
+	if ok, err = d.isVCH(d.appliance); err != nil {
```
Comment (hickeng) [inline lib/install/management/finder.go]: doubt it matters, but good practice would be to only assign to `d.appliance` after we've confirmed that it's a VCH.

File: cmd/vic-machine/configure/configure.go [hunk 12]
```diff
@@ -405,8 +387,19 @@ func (c *Configure) Run(clic *cli.Context) (err error) {
...
+	// (--mem, --cpu, etc) supplied by the user.
+	inputResources := c.Data.ResourceLimits
+	mergedResources := oldData.ResourceLimits
+
+	// overwriting user input w/merged dataset
 	c.Data = oldData
 
+	// Set the ResourceLimits to the input received from
+	// the user
+	oldData.ResourceLimits = inputResources
```
[the remark this answers is not in the document]
Comment (cgtexmex) [reply cmd/vic-machine/configure/configure.go]: The combined resource config set is user input and existing data...doesn't really make sense to validate the existing config...we just retrieved it.  Also the validate code for DRS Disabled warn's when you pass a VCH Pool modification -- specifically that it will be ignored since there are no pools in a DRS disabled environment.  We only want to warn for user provided values and not the existing configuration.

File: cmd/vic-machine/upgrade/upgrade.go [hunk 13]
```diff
@@ -214,9 +215,10 @@ func (u *Upgrade) Run(clic *cli.Context) (err error) {
 	}
 
 	if !u.Data.Rollback {
-		err = executor.Configure(vch, vchConfig, vConfig, false)
+		err = executor.Configure(vchConfig, vConfig)
 	} else {
-		err = executor.Rollback(vch, vchConfig, vConfig)
+		executor.Action = management.RollbackAction
```
[the remark this answers is not in the document]
Comment (cgtexmex) [reply cmd/vic-machine/upgrade/upgrade.go]: agreed

File: lib/install/management/configure.go (hunk 3 above)
Comment (cgtexmex) [reply lib/install/management/configure.go]: We do -- if DRS is enabled and we have an error we will roll back...if DRS is disabled we won't attempt a ResourceConfig since it's not valid.

File: lib/install/management/configure.go (hunk 4 above)
Comment (cgtexmex) [reply lib/install/management/configure.go]: Yeah...we could...this was a quick change this AM on flight to investigate an issue...with this we don't make a property collector as it determines the name off the path..

ultimately `vic-machine` needs to get _most_ if not all the data it needs on init...currently it gets a lot of data multiple times.  So here I used the `.Name()` to avoid yet another round trip...

File: lib/install/management/configure.go (hunk 7 above)
Comment (cgtexmex) [reply lib/install/management/configure.go]: No - we don't store the folder name as part of config (unlike RP) so on portlayer init we grab the current folder...whatever it might be...

File: lib/install/management/configure.go (hunk 8 above)
Comment (cgtexmex) [reply lib/install/management/configure.go]: Possibly...I thought about just using the `tasks.WaitForResult` and adding `SystemError` -- this was older code so _assumed_ that if that was needed it _might_ have surfaced in testing or user experience...happy to change if you think it's worthwhile...

I just glanced at the API documentation around the task...and I think it would be best to use `tasks.WaitForResult` as it should better protect us...

File: lib/install/management/finder.go (hunk 11 above)
Comment (cgtexmex) [reply lib/install/management/finder.go]: agree..but agree in this case it wouldn't matter...

File: lib/install/management/dispatcher.go (hunk 10 above)
Comment (cgtexmex) [reply lib/install/management/dispatcher.go]: No - create was the only call that provided a `conf` -- so the only path that initialized it.  All other calls passed a `nil`

File: lib/install/management/configure.go (hunk 3 above)
Comment (cgtexmex) [reply lib/install/management/configure.go]: Oh wait..I think I misunderstood your question...you mean the rollback sub command?...tricky little devil...

If it's really a rollback I'd say yes...but currently we don't store that anywhere so we'd need to add that...I assume the previous thought was they could `configure` if they didn't like the `pool` settings...

File: lib/install/management/configure.go (hunk 8 above)
Comment (hickeng) [reply lib/install/management/configure.go]: FYI - @AngieCris has added an isTransientError or similar as part of the API list work that just merged. If you chose to update this you may find that useful.

File: lib/install/management/appliance.go [hunk 14]
```diff
@@ -490,21 +485,31 @@ func (d *Dispatcher) createAppliance(conf *config.VirtualContainerHostConfigSpec
...
-			return d.vchVapp.CreateChildVM(ctx, *spec, d.session.Host)
-		})
-	} else {
-		// if vapp is not created, fall back to create VM under default resource pool
-		info, err = tasks.WaitForResult(d.op, func(ctx context.Context) (tasks.Task, error) {
-			return d.session.VMFolder.CreateVM(ctx, *spec, d.vchPool, d.session.Host)
-		})
+	// Create the VCH inventory folder
+	if d.isVC {
+		folderPath := fmt.Sprintf("%s", path.Join(d.session.VMFolder.InventoryPath, conf.Name))
```
Comment (vburenin) [inline lib/install/management/appliance.go]: Why do you need fmt.Sprintf here? seems like not necessary at all.

File: lib/install/management/appliance.go [hunk 15]
```diff
@@ -490,21 +485,31 @@ func (d *Dispatcher) createAppliance(conf *config.VirtualContainerHostConfigSpec
...
+
+		d.op.Info("Creating the VCH folder")
+		// update the session pointer with the VCH Folder
+		d.session.VCHFolder, err = d.session.VMFolder.CreateFolder(d.op, spec.Name)
+		if err != nil {
+			if soap.IsSoapFault(err) {
+				switch soap.ToSoapFault(err).VimFault().(type) {
+				case types.DuplicateName:
```
Comment (vburenin) [inline lib/install/management/appliance.go]: Does the "an existing object with the same name" has any meaning for the end user? What kind of the object? Can we make it a little friendlier?

Comment (vburenin) [reply lib/install/management/appliance.go]: e.g. "The specified VCH name is already in use."

File: lib/install/management/appliance.go [hunk 16]
```diff
@@ -1180,3 +1185,28 @@ func (d *Dispatcher) CheckServiceReady(ctx context.Context, conf *config.Virtual
...
+func (d *Dispatcher) deleteVCHFolder() {
+	// only continue if VC and the VCH Folder is NOT the datacenter wide VM Folder
+	if d.isVC && d.session.VCHFolder != nil && d.session.VCHFolder.Reference() != d.session.VMFolder.Reference() {
+		children, err := d.session.VCHFolder.Children(d.op)
+		if err != nil {
+			d.op.Errorf("Unable to retrieve VCH Folder(%s) contents: %s", d.session.VCHFolder.InventoryPath, err)
+			return
+		}
+		if len(children) > 0 {
+			d.op.Warnf("VCH Folder(%s) contains %d object(s) and will not be removed", d.session.VCHFolder.InventoryPath, len(children))
```
Comment (vburenin) [inline lib/install/management/appliance.go]:
```VCH Folder is not empty and will not be removed."""

File: lib/install/management/create_test.go [hunk 17]
```diff
@@ -17,10 +17,14 @@ package management
 import (
 	"context"
 	"net/url"
+	"path"
 	"testing"
 
 	log "github.com/Sirupsen/logrus"
 
+	"github.com/stretchr/testify/require"
```
Comment (vburenin) [inline lib/install/management/create_test.go]: make one group with logrus?

File: lib/install/management/appliance.go (hunk 15 above)
Comment (cgtexmex) [reply lib/install/management/appliance.go]: Nicely done...I like that...thanks!

File: lib/install/management/appliance.go (hunk 14 above)
Comment (cgtexmex) [reply lib/install/management/appliance.go]: I _think_ that was from a previous tenant...anyhoo...I've removed it...provides little to no value...

File: cmd/vic-machine/configure/configure.go (hunk 12 above)
[the remark this answers is not in the document]
Comment (zjs) [reply cmd/vic-machine/configure/configure.go]: Slightly off-topic: I have a vaguely similar issue in #7797 where validating _just_ the merged configuration doesn't give the desired result (since at that point we can no longer tell whether an existing entity was something _we_ created or whether it's something that a vSphere user created that we don't want to clobber).

I suspect at some point we're going to need to overhaul configure to operate on `(old, delta, new)` instead of assuming that it can just merge everything up-front and operate on that.

File: cmd/vic-machine/upgrade/upgrade.go (hunk 13 above)
[the remark this answers is not in the document]
Comment (zjs) [reply cmd/vic-machine/upgrade/upgrade.go]: :+1:; I think it's going to have to be a separate action in the API

File: lib/install/management/configure.go (hunk 3 above)
Comment (zjs) [reply lib/install/management/configure.go]: Should we open an issue to track this? (Note to self: this will _also_ be a concern for #7797.)

Rolling back non-settings changes is... hard.

File: lib/install/management/inventory.go [hunk 18]
```diff
@@ -0,0 +1,217 @@
...
+	refs, err := compute.VM(d.op, d.session, pool)
+	if err != nil {
+		err = fmt.Errorf("Error retrieving vms: %s", err)
+		return nil, err
+	}
+
+	return vm.Attributes(d.op, d.session, refs)
+}
+
+func (d *Dispatcher) isContainer(vm mo.VirtualMachine) bool {
```
Comment (zjs) [inline lib/install/management/inventory.go]: How does this relate to `isContainerVM`?

https://github.com/vmware/vic/blob/0ff435cd277c44b90b36f93b837c1c45d47500b3/lib/install/management/appliance.go#L95-L111

File: lib/install/validate/config_to_data.go (hunk 1 above)
Comment (zjs) [reply lib/install/validate/config_to_data.go]: What about `config.hardware.numCPU` and `config.hardware.memoryMB` (used in the function above)?

File: lib/install/management/inventory.go (hunk 18 above)
Comment (cgtexmex) [reply lib/install/management/inventory.go]: It will eventually replace it...instead of doing that in this PR (that's already huge) I was waiting to break it out into chunks -- I planned on documenting what I think we should do and then discuss with you, George, etc..

Basically we need to rework the interactions w/VC as vic-machine is very chatty...for example when you delete it will pull the config 2x for each VM...I determined the other day on a `vic-machine delete` there are roughly 9 calls to VC for each container it's deleting...so have 10 containers that's 90 calls...and that's just a portion of the interaction...we need to do a much better job of protecting the infrastructure...

File: lib/install/validate/config_to_data.go (hunk 1 above)
Comment (cgtexmex) [reply lib/install/validate/config_to_data.go]: I actually thought I switched this back...but regardless this is returning the [resourceConfigSpec](https://vdc-download.vmware.com/vmwb-repository/dcr-public/6b586ed2-655c-49d9-9029-bc416323cb22/fa0b429a-a695-4c11-b7d2-2cbc284049dc/doc/index.html#linkf89ddfb4da9f5fead1008dc6689b547fb425dc2b;vim.ResourcePool.html) which contains the two data objects we were returning and would now include a string, moref & datetime...

File: lib/install/management/configure.go [hunk 19]
```diff
@@ -98,50 +98,63 @@ func (d *Dispatcher) Configure(vch *vm.VirtualMachine, conf *config.VirtualConta
...
-		d.retryDeleteSnapshotByRef(newSnapshotRef, snapshotName, conf.Name)
+	}
 
-		// return the error message for upgrade
-		return err
+	// If successful try to grant permissions to the ops-user
+	if conf.ShouldGrantPerms() {
+		err := opsuser.GrantOpsUserPerms(d.op, d.session, conf)
+		if err != nil {
+			err = errors.Errorf("Failed to grant permissions to ops-user, failure: %s", err)
```
Comment (jzt) [inline lib/install/management/configure.go]: Should this be `return errors.Errorf(...)`? Otherwise it seems to be ignored.

Comment (cgtexmex) [reply lib/install/management/configure.go]: hmmm...that seems problematic...i.e. needs to be changed...

File: cmd/vic-machine/configure/configure.go (hunk 12 above)
[the remark this answers is not in the document]
Comment (zjs) [reply cmd/vic-machine/configure/configure.go]: Also off-topic, but I'm leaving it here so that I can find it again: We seem to have another case where operating on just the merged configuration doesn't work well and work around it in a (different) one-off way:

https://github.com/vmware/vic/blob/7fd77b0f05a9c3227d461f983261784e6cc6f381/lib/install/management/configure.go#L168-L177

File: lib/install/management/configure.go [hunk 20]
```diff
@@ -98,50 +98,65 @@ func (d *Dispatcher) Configure(vch *vm.VirtualMachine, conf *config.VirtualConta
...
 	newSnapshotRef, err := d.createSnapshot(snapshotName, "configure snapshot")
 	if err != nil {
-		d.deleteUpgradeImages(ds, settings)
+		d.deleteISOs(ds, settings)
 		return err
 	}
 
-	err = d.update(conf, settings, isConfigureOp)
+	// rollback function
+	rollback := func() {
```
Comment (zjs) [inline lib/install/management/configure.go]: I'm not sure I love this rollback function pattern. I don't love the pattern that was here before either, but this is still really fragmented. For example, we still have a one-off `defer` above to do one piece of the rollback and now it seems like we leave the snapshot behind when we rollback (instead of rolling back and then deleting it) since we `return err` after `rollback()`.
